src/metabase/upload.clj

```diff
@@ -633,67 +637,74 @@
+            old-column-types   (map (comp base-type->upload-type :base_type normed-name->field) normed-header)
+            ;; in the happy, and most common, case all the values will match the existing types
+            ;; for now we just plan for the worst and perform a fairly expensive operation to detect any type changes
+            ;; we can come back and optimize this to an optimistic-with-fallback approach later.
+            detected-types     (column-types-from-rows settings old-column-types rows)
+            new-column-types   (map #(if (matching-or-upgradable? %1 %2) %2 %1) old-column-types detected-types)
+            _                  (when (and (not= old-column-types new-column-types)
+                                          ;; if we cannot coerce all the columns, don't bother coercing any of them
+                                          ;; we will instead throw an error when we try to parse as the old type
+                                          (= detected-types new-column-types))
+                                 (let [fields (map normed-name->field normed-header)]
+                                   (->> (changed-field->new-type fields old-column-types detected-types)
+                                        (alter-columns! driver database table))))
+            ;; this will fail if any of our required relaxations were rejected.
+            parsed-rows        (parse-rows settings new-column-types rows)
+            row-count          (count parsed-rows)
+            stats              {:num-rows          row-count
+                                :num-columns       (count new-column-types)
+                                :generated-columns (if create-auto-pk? 1 0)
+                                :size-mb           (file-size-mb file)
+                                :upload-seconds    (since-ms timer)}]
+
+        (try
+          (driver/insert-into! driver (:id database) (table-identifier table) normed-header parsed-rows)
+          (catch Throwable e
+            (throw (ex-info (ex-message e) {:status-code 422}))))
+
+        (when create-auto-pk?
+          (driver/add-columns! driver
+                               (:id database)
+                               (table-identifier table)
+                               {auto-pk-column-keyword (driver/upload-type->database-type driver ::auto-incrementing-int-pk)}
+                               :primary-key [auto-pk-column-keyword]))
+
+        (scan-and-sync-table! database table)
+
+        (when create-auto-pk?
+          (let [auto-pk-field (table-id->auto-pk-column (:id table))]
+            (t2/update! :model/Field (:id auto-pk-field) {:display_name (:name auto-pk-field)})))
+
+        (events/publish-event! :event/upload-append
+                               {:user-id  (:id @api/*current-user*)
+                                :model-id (:id table)
+                                :model    :model/Table
```

Comment:
If we want to track the model_id that was uploaded from, that is only available on the frontend.

Reply:
Feel like I'm missing something in your comment, so apologies if my response doesn't quite speak to it.

We use the underlying table itself as the anchor resource for the event, not the model - this is consistent with the event published on initial upload. For the initial upload the `model_id` for the model that we create is put inside the `:details` map instead.

Using the Model whose page we append from on the frontend is not something I thought of, slightly different but compatible semantics. It still leaves us without a natural id to include if the action was taken from the Collections screen though, so I don't see the harm in just excluding it from both for now.

Reply:
resolved on zoom